### ISSUE TITLE
`Style/RedundantBegin` should detect redundant begin around memoization

### ISSUE THREAD

**Issue by deivid-rodriguez** [rubocop/rubocop#9601]
## Is your feature request related to a problem? Please describe.

I'd like `rubocop` to be able to detect redundant begin/end blocks around memoization, like here: https://github.com/rubocop/rubocop/commit/3b48d8bdf5b1c2e05e35061837309890f04ab08c.

## Describe the solution you'd like

I think it would fit nicely inside the existing `Style/RedundantBegin` cop.

## Describe alternatives you've considered

A separate cop if excluding this case was intentional, I guess.

**Comment by koic** [comment]
Good catch! I think it can enhance `Style/RedundantBegin` cop.

**Comment by deivid-rodriguez** [comment]
Thanks!!

**Comment by koic** [comment]
@deivid-rodriguez You are welcome! 😄

**Comment by tas50** [comment]
Thanks @deivid-rodriguez and @koic https://github.com/chef/chef/pull/11220

**Comment by Drenmi** [comment]
Would have been nice to have a configuration option, as we deliberately use this style for complex memoization expressions. 😬

**Comment by deivid-rodriguez** [comment]
@Drenmi Can you show an example of how this fix complicates indentation, if that's what you mean by indent-mega-blocks? Isn't

```ruby
def foo
  @foo ||= begin
    mega_block do
      # ...      
      # a lot of code
      # ...
    end
  end
end 
```

equivalent to

```ruby
def foo
  @foo ||=
    mega_block do
      # ...
      # a lot of code
      # ...
    end
end 
```

in terms of indentation? Or does it violate another common style rule?

**Comment by Drenmi** [comment]
The auto-correction with our current settings leaves:

```ruby
def something
  @something ||= mega_block do
                   # ...
                 end
end
```

after removing `begin` and applying other rules. Another small point that caused confusion/mistakes in the past is the mismatched `end` when not using `begin`.

**Comment by deivid-rodriguez** [comment]
Ah I see. Yeah, I've been there. I don't like this mega-indented blocks and I do recall having to change the code manually to the alternative I mentioned since rubocop auto-correct prefers the other style. I don't get the problem with mismatched `end` though, where is the mismatch?

**Comment by alexventuraio** [comment]
Kind of the same issue from @Drenmi here:
```ruby
# before
def decorator
  @decorator ||= begin
    case reminder.remindable_type
    when 'EmailSchedule'
      Notifier::EmailScheduleDecorator.new(reminder)
    when 'ReportScheduled'
      Notifier::ReportScheduledDecorator.new(reminder)
    else
      raise 'Unsupported type of Notifie'
    end
  end
end

# after
def decorator
  @decorator ||= case reminder.remindable_type
                  when 'EmailSchedule'
                    Notifier::EmailScheduleDecorator.new(reminder)
                  when 'ReportScheduled'
                    Notifier::ReportScheduledDecorator.new(reminder)
                  else
                    raise 'Unsupported type of Notifier'
                  end
end

```
